indexing-service/src/main/java/org/apache/druid/indexing/common/task/KillUnusedSegmentsTask.java

```diff
@@ -128,6 +146,9 @@ public TaskStatus runTask(TaskToolbox toolbox) throws Exception
     }
 
     // Kill segments
+    unusedSegments = maxSegmentsToKill == null
+        ? unusedSegments
```

Comment:
Instead of pruning out the unused segments up here, have we considered pushing down the limit to the [metadata](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java#L119) query so we ask only what we want? I think that should considerably reduce the load on the metadata store, reduce the lock duration, etc, for unused segments that aren't killed in a run anyway. What do you think?

Reply:
I thought it was best to mark all segements in interval specified as unused. Seemed less expensive than actually deleting the segment from deep storage, and other things involved in killing the segment. I did consider this though. But with the limit also applied to marking unused. A user might need multiple requests to mark segments in interval as unused, where as if the limit is only applied to kill, then all segments in interval are marked unused, and if we hit limit on kill, auto kill will clean up eventually, if enabled. Let me know what you think.

Reply:
I think the comment was referring to something different, about adjusting `RetrieveUnusedSegmentsAction(getDataSource(), getInterval()))` so that it can accept a limit and return a smaller set of unused segments

Reply:
yes! sorry, I misunderstood, fixing now

Reply:
fixed

Reply:
IMO, kill tasks shouldn't mark stuff unused at all. It's best if there is a time delay between markings things unused and permanently deleting them. It's possible for people to make mistakes with what they delete, and we want the mistakes to be undoable for some period of time.

I think the best thing to do here is actually deprecate the `markUnused` parameter, and focus on workflows that are like:

- Users delete stuff by marking it unused
- Later, the system deletes it permanently by way of autokill (which runs these `kill` tasks)

Btw, an important part of making this work is including a buffer period between when a segment is marked unused, and when autokill would remove it. This was discussed in #12526 although I am not sure what the current status is.

For _this_ patch, IMO the best approach is to have the Coordinator send a list of segments to the `kill` task, rather than `interval` + `maxSegmentsToKill`. This will make it easier to implement the buffer period we need for autokill later— the Coordinator would be in charge of determining which specific segments are killable, and it would need the `kill` task to respect that specific list. This also side-steps the issue with `markAsUnused`, since the `maxSegmentsToKill` parameter would no longer exist.

Reply:
Thanks for your input @gianm . This makes sense to me. I guess kill tasks will have another config property added, like `segmentIds` , and if that is set, the killTask will ignore the other config properties like `interval`  and `datasource` ?

Reply:
Added a new task type to kill unused segments by id.